apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,242 @@
+import {
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+  Logger,
+  ConflictException,
+  HttpException,
+  InternalServerErrorException,
+  ServiceUnavailableException,
+} from '@nestjs/common';
+import { CacheService } from '@novu/application-generic';
+import { Observable, of, throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+import { createHash } from 'crypto';
+import * as jwt from 'jsonwebtoken';
+import { IJwtPayload } from '@novu/shared';
+
+const LOG_CONTEXT = 'IdempotencyInterceptor';
+const IDEMPOTENCY_CACHE_TTL = 60 * 60 * 24; //24h
+const IDEMPOTENCY_PROGRESS_TTL = 60 * 5; //5min
+
+const HEADER_KEYS = {
+  IDEMPOTENCY_KEY: 'idempotency-key',
+  RETRY_AFTER: 'retry-after',
+  IDEMPOTENCY_CONFLICT: 'x-idempotency-conflict',
+  IDEMPOTENCY_REPLAYED: 'idempotency-replayed',
+};
+
+enum IdempotencyConflictEnum {
+  REQUEST_PROCESSING = 'REQUEST_PROCESSING',
+  DIFFERENT_BODY = 'DIFFERENT_BODY',
+}
+
+enum ReqStatusEnum {
+  PROGRESS = 'in-progress',
+  SUCCESS = 'success',
+  ERROR = 'error',
+}
+
+@Injectable()
+export class IdempotencyInterceptor implements NestInterceptor {
+  constructor(private readonly cacheService: CacheService) {}
+
+  async intercept(context: ExecutionContext, next: CallHandler): Promise<Observable<any>> {
+    const request = context.switchToHttp().getRequest();
+    const idempotencyKey = this.getIdempotencyKey(context);
+    const isEnabled = process.env.API_IDEMPOTENCY_ENABLED == 'true';
+    if (!isEnabled || !idempotencyKey || !['post', 'patch'].includes(request.method.toLowerCase())) {
+      return next.handle();
+    }
+    const cacheKey = this.getCacheKey(context);
+
+    try {
+      const bodyHash = this.hashRequestBody(request.body);
+      //if 1st time we are seeing the request, marks the request as in-progress if not, does nothing
+      const isNewReq = await this.setCache(
+        cacheKey,
+        { status: ReqStatusEnum.PROGRESS, bodyHash },
+        IDEMPOTENCY_PROGRESS_TTL,
+        true
+      );
```

Comment:
@rifont  here,
for: https://github.com/novuhq/novu/pull/4531#discussion_r1360768330